elki-outlier/src/main/java/elki/outlier/density/HySortOD.java

```diff
@@ -0,0 +1,743 @@
+
+  /**
+   * Number of bins
+   */
+  protected int b;
+
+  /**
+   * Threshold to balance the trade-off between the number of hyper-cubes to
+   * scan
+   * during the search and the granularity of mapping
+   */
+  protected int minSplit;
+
+  // hypercube's length
+  private final double l;
+
+  // to match each object with its hyper-cube
+  // it's a little trick to deal with hypercube rearrangement
+  private int[] invertedIndex;
+
+  // strategy for hyper-cubes density search
+  private DensityStrategy strategy;
+
+  /**
+   * Constructor
+   * 
+   * @param k
+   * @param b Number of bins
+   */
+  public HySortOD(int b, int minSplit) {
+    super();
+    this.b = b;
+    this.l = 1 / (double) this.b;
+    this.minSplit = minSplit;
+    if(this.minSplit > 0) 
+      this.strategy = new TreeStrategy(this.minSplit);
+     else 
+      this.strategy = new NaiveStrategy();
+  }
+
+  public OutlierResult run(Database db, Relation<V> relation) {
+
+    // Output data storage
+    WritableDoubleDataStore scores = DataStoreUtil.makeDoubleStorage(relation.getDBIDs(), DataStoreFactory.HINT_DB);
+    FiniteProgress prog = LOG.isVerbose() ? new FiniteProgress("HySortOD scores", relation.size(), LOG) : null;
+    // Track minimum and maximum scores
+    DoubleMinMax minmax = new DoubleMinMax();
+
+    int[] W;
+    List<Hypercube> H;
+
+    {
+      H = getSortedHypercubes(relation);
+      W = this.strategy.buildIndex(H).getDensities();
+
+      // Iterate over all objects
+      int objectPosition = 0;
+      for(DBIDIter iter = relation.iterDBIDs(); iter.valid(); iter.advance())
+        scores.putDouble(iter, score(W[this.invertedIndex[objectPosition++]]));
+
+      LOG.incrementProcessed(prog);
+    }
+
+    LOG.ensureCompleted(prog);
+    // Wrap the result in the standard containers
+    OutlierScoreMeta meta = new BasicOutlierScoreMeta(
+        // Actually observed minimum and maximum values
+        minmax.getMin(), minmax.getMax(),
+        // Theoretical minimum and maximum: no variance to infinite variance
+        0, Double.POSITIVE_INFINITY);
+    DoubleRelation rel = new MaterializedDoubleRelation("HySortOD", relation.getDBIDs(), scores);
+    return new OutlierResult(meta, rel);
+  }
+
+  private List<Hypercube> getSortedHypercubes(Relation<V> relation) {
+    BinaryTree<Hypercube> sorted = new BinaryTree<>();
+
+    int objectPosition = 0;
+    // Iterate over all objects
+    for(DBIDIter iter = relation.iterDBIDs(); iter.valid(); iter.advance()) {
+      Hypercube h = new Hypercube(relation.get(iter), this.l);
+      Hypercube obj = sorted.get(h);
+
+      if(Objects.isNull(obj)) {
+        h.add(objectPosition);
+        sorted.add(h);
+      }
+      else {
+        obj.add(objectPosition);
+      }
+
+      objectPosition++;
+    }
+
+    int n = sorted.size();
+    List<Hypercube> H = new ArrayList<Hypercube>(n);
+
+    this.invertedIndex = new int[relation.size()];
```

Comment:
If you used ELKIs DBID data structures above, you would not need this inverted lookup.